0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+  /// from the bivariate polynomial are valid.
+  ///
+  /// This method can be called over an insecure channel as the matrix
+  /// doesn't contain any sensitive information.
+  fn verification_matrix(&self, churp_id: u8, round: u64);
+
+  /// Return bivariate share (polynomial) for the calling node.
+  /// 
+  /// The polynomial is partial evaluation of the bivariate polynomial
+  /// at a given x or y value (caller id).
+  /// 
+  /// Polynomial:
+  ///     g_{node_id}(y) = B(caller_id,y) (dealing phase)
+  ///     g_{node_id}(x) = B(x,caller_id) (proactive randomization)
+  ///
+  /// Bivariate polynomial:
+  ///     B(x,y) = \sum_{i=0}^{t_n} \sum_{j=0}^{t_m} a_{i,j} x^i y^j
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn bivariate_share(&self, churp_id: u8, round: u64);
+
+  /// Construct a full share from the verified bivariate shares obtained
+  /// from the dealers.
+  ///
+  /// Full secret:
+  ///      s(y) =  B(node_id,y) = \sum_g_{node_id}(y)
+  fn dealing(&self, churp_id: u8, round: u64);
+
+  /// Construct a reduced share from the switch data points obtained from
+  /// the nodes in the current committee.
+  ///
+  /// Reduced share:
+  ///      r(x) = B(x, node_id)
+  fn share_reduction(&self, churp_id: u8, round: u64);
+
+  /// Randomize reduced share using bivariate shares obtained from
+  /// the nodes in the new committee.
+  ///
+  /// Proactive reduced share:
+  ///      r'(x) = B(x, node_id) + \sum Q(x, node_id)
+  fn proactive_randomization(&self, churp_id: u8, round: u64);
+
+  /// Construct a full share from the switch data points obtained from 
+  /// the nodes in the new committee.
+  ///
+  /// Full share:
+  ///      s'(y) = B'(node_id,y)
+  fn full_share_distribution(&self, churp_id: u8, round: u64);
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId) -> Integer;
```

Comment:
Instead of just a `KeyPairId` the caller should need to provide some sort of proof that key manager nodes/enclaves can independently verify before releasing the share to the caller.

Reply:
I think that the signature of this method is ok as it is. However, this method should have a midlleware that handles access control. And that could be defined in another adr and added to the CHURP latter.